keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,299 @@
+* Scheduler will respect `Node.status.allocatable` and `Node.status.capacity` for CSI volumes if `CSINode` object is not available or has missing entry in `CSINode.spec.drivers[xyz].allocatable` during a deprecation period but kubelet will stop populating `Node.status.allocatable` and `Node.status.capacity` for CSI volumes.
+  * After deprecation period for CSI volumes, limits coming from `Node.status.allocatable` and `Node.status.capacity` will be completely ignored.
+  * After deprecation period, scheduler won't schedule any pods that use CSI volumes to a node with missing `CSINode` instance. It is expected that it happens only during node registration when `Node` exists and `CSINode` doesn't and it self-heals quickly.
+  * After deprecation period, if `CSINode` exists and  does not contain a CSI driver, scheduler will not put CSI volumes for the driver to the node.
+
+
+#### Implementation detail for in-tree Drivers with CSI migration disabled
+
+##### When no CSI driver for same underlying storage type is installed on the node.
+
+* For Azure, GCEPD, AWS and Cinder - in-tree volume plugins will keep reporting their limits via `Node` object and old predicates will work as expected until CSI migration has been enabled (and GA) for given volume plugin.
+
+##### When CSI driver for same underlying storage type is installed on the node.
+
+* For Azure, GCEPD, AWS and Cinder - in-tree volume plugins will report their limits via `Node` object same as before.
+* Old predicates will be modified to perform an additional check of `CSINode` at very beginning of their code. If they detect that `CSINode` object contains limit of same underlying volume type, they will return early(with success) and `MaxCSIVolumeCountPred` will be responsible for counting both CSI and in-tree volumes of same type.
+
+
+#### Implementation detail for in-tree Drivers with CSI migration enabled
+
+* For Azure, GCEPD, AWS and Cinder - in-tree volume plugins will report their limits via `Node` object same as before.
+* As described in section above - old predicates will be modified to perform an additional check of `CSINode` object at very beginning of their code. If they detect that `CSINode` object contains limit of same underlying volume type or CSI migration has been enabled for the volume, the old predicate will return early(with success) and `MaxCSIVolumeCountPred` will be responsible for counting both CSI and in-tree volumes of same type.
+* `MaxCSIVolumeCountPred` will be updated to prevent scheduling of pods with volumes(in-tree or CSI) for which CSI migration has been enabled and driver is not installed yet. Please note this is different from other CSI drivers(for which there is no in-tree to CSI migration plan) where we will follow a deprecation cycle before preventing pod scheduling for unavailable CSI drivers.
```

Comment:
Maybe add one more note that when CSI migration for that plugin is GA (including kubelet version skew) and cannot be disabled, the old predicate for that plugin can be removed from the scheduler.

Reply:
It is covered in goals section:

>Existing predicates for in-tree volumes `MaxEBSVolumeCount`, `MaxGCEPDVolumeCount`, `MaxAzureDiskVolumeCount` and `MaxCinderVolumeCount`(now deprecated) will be removed when in-tree to CSI migration is GA and enabled by default for that particular volume plugin.